File: LeetCode 33 - Search in Rotated Sorted Array.java
// Time Complexity: Since we use binary search methods here, 
// so the time complexity will be O(logn).
/* --------------------------------------------------------*/
// Space Complexity: the space complexity will be O(1).

class Solution {
    public int search(int[] nums, int target) {
        if(nums == null || nums.length == 0)
            return -1;
        int low = 0, high = nums.length - 1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                return mid;
            }else{
                // check whether left sorted
                if(nums[mid] >= nums[low]){
                    // check which interval the target locates.
                    if(nums[low] <= target && nums[mid] > target){
                        high = mid -1;
                    }else{
                        low = mid + 1;
                    }
                }
                // check whether right sorted
                else{
                    // check which interval the target locates.
                    if(nums[high] >= target && nums[mid] < target){
                        low = mid + 1;
                    }else{
                        high = mid - 1;
                    }
                }
            } 
        }
        return  -1;
    }
}

File: Leetcode-33_Search_in_Rotated_Sorted_Array.cpp
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, initially on deciding search space.

class Solution {
public:
    int search(vector<int>& nums, int target) 
    {
        int low =0;
        int high =nums.size()-1;
        while(low<=high)
        {
            int mid = low+(high-low)/2;
            
            if(nums[mid]==target)
                return mid;
            
            if(nums[low]<=nums[mid]) // left half is sorted
            {
                if(target>= nums[low] && nums[mid] > target) // target between low and high
                    high=mid-1;
                else
                    low = mid+1; // ele in unsorted half
            }
            
            else // right half is sorted
            {
                if(nums[high] >= target && target > nums[mid]) // target between low and high
                    low = mid+1;
                else
                    high = mid-1; // ele in unsorted half
            }
                  
        }
        
        return -1;
    }
};